agent/ebs/watcher.go

```diff
@@ -0,0 +1,175 @@
+				w.NotifyFound(foundVolumes)
+			}
+		case <-w.ctx.Done():
+			w.scanTicker.Stop()
+			log.Info("EBS Watcher Stopped due to agent stop")
+			return
+		}
+	}
+}
+
+// Stop will stop the EBS watcher
+func (w *EBSWatcher) Stop() {
+	log.Info("Stopping EBS watcher.")
+	w.cancel()
+}
+
+// HandleResourceAttachment processes the resource attachment message. It will:
+// 1. Check whether we already have this attachment in state and if so it's a noop.
+// 2. Otherwise add the attachment to state, start its ack timer, and save to the agent state.
+func (w *EBSWatcher) HandleResourceAttachment(ebs *apiebs.ResourceAttachment) error {
+	attachmentType := ebs.GetAttachmentProperties(apiebs.ResourceTypeName)
+	if attachmentType != apiebs.ElasticBlockStorage {
+		log.Warnf("Resource type not Elastic Block Storage. Skip handling resource attachment with type: %v.", attachmentType)
+		return nil
+	}
+
+	volumeId := ebs.GetAttachmentProperties(apiebs.VolumeIdName)
+	_, ok := w.agentState.GetEBSByVolumeId(volumeId)
+	if ok {
+		log.Infof("EBS Volume attachment already exists. Skip handling EBS attachment %v.", ebs.EBSToString())
+		return nil
+	}
+
+	if err := w.addEBSAttachmentToState(ebs); err != nil {
+		return fmt.Errorf("%w; attach %v message handler: unable to add ebs attachment to engine state: %v",
+			err, attachmentType, ebs.EBSToString())
+	}
+
+	return nil
+}
+
+// NotifyFound will go through the list of found EBS volumes from the scanning process and mark them as found.
+func (w *EBSWatcher) NotifyFound(foundVolumes []string) {
+	for _, volumeId := range foundVolumes {
+		w.notifyFoundEBS(volumeId)
+	}
+}
+
+// notifyFoundEBS will mark it as found within the agent state
+func (w *EBSWatcher) notifyFoundEBS(volumeId string) {
+	// TODO: Add the EBS volume to data client
+	ebs, ok := w.agentState.GetEBSByVolumeId(volumeId)
+	if !ok {
+		log.Warnf("Unable to find EBS volume with volume ID: %v.", volumeId)
+		return
+	}
+
+	if ebs.HasExpired() {
+		log.Warnf("EBS status expired, no longer tracking EBS volume: %v.", ebs.EBSToString())
+		return
+	}
+
+	if ebs.IsSent() {
+		log.Warnf("State change event has already been emitted for EBS volume: %v.", ebs.EBSToString())
+		return
+	}
+
+	if ebs.IsAttached() {
+		log.Infof("EBS volume: %v, has been found already.", ebs.EBSToString())
+		return
+	}
+
+	ebs.StopAckTimer()
+	ebs.SetAttachedStatus()
+
+	log.Infof("Successfully found attached EBS volume: %v", ebs.EBSToString())
+}
+
+// removeEBSAttachment removes a EBS volume with a specific volume ID
+func (w *EBSWatcher) removeEBSAttachment(volumeID string) {
+	// TODO: Remove the EBS volume from the data client.
+	w.agentState.RemoveEBSAttachment(volumeID)
+}
+
+// addEBSAttachmentToState adds an EBS attachment to state, and start its ack timer
+func (w *EBSWatcher) addEBSAttachmentToState(ebs *apiebs.ResourceAttachment) error {
+	volumeId := ebs.AttachmentProperties[apiebs.VolumeIdName]
+	err := ebs.StartTimer(func() {
```

Comment:
What is the behavior of this timer if Agent restarts before when the timer is running?

Reply:
With the current implementation, the timer won't stop which is similar to what we have with ENI attachments IIUC. -> https://github.com/aws/amazon-ecs-agent/blob/master/agent/acs/handler/attach_eni_handler_common.go#L80C21-L80C21

Although we do reset the timer when we get another ENI attachment message. Perhaps we can follow the same behavior with EBS attachments.

Reply:
Ah, we'd probably also want the same behavior if the same attachment message was sent over. We start the ACK timer if it hasn't done so yet. There's a check in `StartTimer` if it hasn't been set yet so this would be idempotent.

Reply:
Thanks @mye956. I see that the timer is set up in `Initialize` method of `ResourceAttachment` too based on the expiration of the attachment loaded from the state file. LGTM!

Reply:
Actually, upon second look it appears that `Initialize` method is currently not called anywhere. Let's make sure that we call it somewhere in Agent startup workflow. This can be done in a future PR. Thanks!

Reply:
Noted, thanks Amogh!